spec/rspec/core/reentrant_mutex_spec.rb

```diff
@@ -0,0 +1,37 @@
+require 'rspec/core/reentrant_mutex'
+
+# There are no assertions specifically
+# They are pass if they don't deadlock
+RSpec.describe RSpec::Core::ReentrantMutex do
+  let!(:mutex) { described_class.new }
+  let!(:order) { ThreadOrderSupport.new }
+  after { order.apocalypse! }
+
+  it 'can repeatedly synchronize within the same thread' do
+    mutex.synchronize { mutex.synchronize { } }
+  end
+
+  it 'locks other threads out while in the synchronize block' do
+    order.declare(:before) { mutex.synchronize { } }
+    order.declare(:within) { mutex.synchronize { } }
+    order.declare(:after)  { mutex.synchronize { } }
+
+    order.pass_to :before, :resume_on => :exit
+    mutex.synchronize { order.pass_to :before, :resume_on => :sleep }
+    order.pass_to :before, :resume_on => :exit
+  end
+
+  it 'resumes the next thread once all its synchronize blocks have completed' do
+    order.declare(:thread) { mutex.synchronize { } }
+    mutex.synchronize { order.pass_to :thread, :resume_on => :sleep }
+    order.apocalypse! :join
+  end
+
+  it 'is implemented without depending on the stdlib' do
+    loaded_filenames = $LOADED_FEATURES.map { |filepath| File.basename filepath }
+    pending 'thread seems to be required from core, and something is still requiring monitor'
+    expect(loaded_filenames).to_not include 'monitor.rb'
+    expect(loaded_filenames).to_not include 'thread.rb'
+    expect(loaded_filenames).to_not include 'thread.bundle'
+  end
```

Comment:
FWIW, we now have library-wide checks for minimizing the parts of stdlib we load (it was added in the last few weeks).  It's defined in rspec-support:

https://github.com/rspec/rspec-support/blob/154c598971338108457329c44d276cd1a64deded/lib/rspec/support/spec/library_wide_checks.rb#L92-L107

That example group gets included here, where we also define the whitelist of allowed stdlibs for rspec-core:

https://github.com/rspec/rspec-core/blob/95609709de8c5df85867feda0b9573b5e53a14bd/spec/rspec/core_spec.rb#L5-L65

So this spec isn't needed.  I'm curious about why thread/monitor are getting loaded...I'll take a look.

On a side note, your 3 expectations can be collapsed into one:

``` ruby
expect(loaded_filenames).to_not include('monitor.rb', /thread/)
```